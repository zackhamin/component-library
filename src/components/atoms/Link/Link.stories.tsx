import { Error } from 'Atoms/Icons';
import { Link } from './Link';
import { ConditionalLinkProps, LinkProps } from './Link.types';
import { Meta, Story } from '@storybook/react';

export default {
	title: 'atoms/Link',
	component: Link,
	argTypes: {
		variant: {
			control: {
				type: 'select',
				options: ['button', 'link'],
			},
		},
		size: {
			control: {
				type: 'select',
				options: ['small', 'medium', 'large'],
			},
		},
		style: {
			control: {
				type: 'select',
				options: ['primary', 'secondary', 'icon'],
			},
		},
		as: {
			control: {
				type: 'select',
				options: ['a'],
			},
		},
		block: {
			control: {
				type: 'boolean',
			},
		},
		disabled: {
			control: {
				type: 'boolean',
			},
		},
	},
} as Meta;

const defaultProps = {
	children: 'Link Text',
} as LinkProps;

const Template: Story<LinkProps & ConditionalLinkProps> = (args) => (
	<Link href="#" {...args} />
);

export const Default = Template.bind({});
Default.args = {
	...defaultProps,
};

export const PrimaryBtn = Template.bind({});
PrimaryBtn.args = {
	...defaultProps,
	variant: 'button',
	style: 'primary',
};

export const SecondaryBtn = Template.bind({});
SecondaryBtn.args = {
	...defaultProps,
	variant: 'button',
	style: 'secondary',
};

export const IconBtn = Template.bind({});
IconBtn.args = {
	...defaultProps,
	variant: 'button',
	style: 'icon',
	children: <Error color="#333333" fontSize="2.2rem" />,
};
