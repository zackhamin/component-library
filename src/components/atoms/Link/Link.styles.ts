import { colors } from 'Styles';
import { SharedStyles } from 'Atoms/Button/Button.styles';
import styled from 'styled-components';

const StyledLink = styled.a`
	&.variant-link {
		font-size: 1.6rem;
		display: inline-block;
		text-align: center;

		&:hover {
			text-decoration: underline;
		}

		&:focus-visible {
			outline: solid 2px ${colors.emerald75};
			outline-offset: 3px;
			border-radius: 1px;
		}
	}

	&.variant-button {
		${SharedStyles}
		font-size: 1.6rem;
		display: inline-flex;
		gap: 1rem;
		justify-content: center;
		text-align: center;
		text-decoration: none;

		&.block {
			width: 100%;
		}

		&[disabled] {
			cursor: not-allowed !important;
			opacity: 0.5;
			flex-grow: 1;
		}
	}
`;

export { StyledLink };
