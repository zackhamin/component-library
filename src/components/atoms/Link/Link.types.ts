import { PolymorphicComponentPropsWithRef } from '@Types/polymorphism-utils';

type LinkButtonProps = {
	variant: 'button';
	style: 'primary' | 'secondary' | 'icon';
	size: 'large' | 'medium' | 'small';
	block?: boolean;
	disabled?: boolean;
};

type DefaultLinkProps = {
	variant: 'link';
};

export interface LinkProps {
	children: React.ReactNode;
	variant?: 'button' | 'link';
}

export type ConditionalLinkProps = LinkButtonProps | DefaultLinkProps;

export type PolymorphicLinkProps<Component extends React.ElementType> =
	PolymorphicComponentPropsWithRef<Component, LinkProps>;
