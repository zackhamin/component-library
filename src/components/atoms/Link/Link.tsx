import classNames from 'classnames';
import { forwardRef } from 'react';
import { PolymorphicRef } from '@Types/polymorphism-utils';
import { StyledLink } from './Link.styles';
import { ConditionalLinkProps, PolymorphicLinkProps } from './Link.types';

const Link = forwardRef(
	<Component extends React.ElementType = 'a'>(
		{
			as,
			children,
			style,
			size,
			block,
			disabled = false,
			variant = 'link',
			...rest
		}: PolymorphicLinkProps<Component> & ConditionalLinkProps,
		ref?: PolymorphicRef<Component>
	) => {
		const tag = as || 'a';

		const classes = classNames(style, {
			[`variant-${variant}`]: variant,
			[`btn-${size}`]: size,
			block,
		});

		return (
			<StyledLink
				ref={ref}
				className={classes}
				as={disabled ? 'span' : tag}
				disabled={disabled}
				{...rest}
			>
				{children}
			</StyledLink>
		);
	}
);

export { Link };
