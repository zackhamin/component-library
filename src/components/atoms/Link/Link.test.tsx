import { Link } from '../Link';
import { render } from '@testing-library/react';

describe('Link', () => {
	it('renders correctly', () => {
		const { container } = render(<Link>Link Text</Link>);
		expect(container).toMatchSnapshot();
	});

	it('renders as a span when disabled as button', () => {
		const { container } = render(
			<Link variant="button" disabled>
				Link Text
			</Link>
		);

		const span = container.querySelector('span');

		expect(span).toBeInTheDocument();
	});

	it('correctly handles polymorphism', () => {
		// Intended to be used with link components - this is just to test polymorphism
		const { container } = render(<Link as="span">Link Text</Link>);

		const span = container.querySelector('span');

		expect(span).toBeInTheDocument();
	});
});
