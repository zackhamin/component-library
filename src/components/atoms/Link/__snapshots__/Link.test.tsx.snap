// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders correctly 1`] = `
.c0.variant-link {
  font-size: 1.6rem;
  display: inline-block;
  text-align: center;
}

.c0.variant-link:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0.variant-link:focus-visible {
  outline: solid 2px #018476;
  outline-offset: 3px;
  border-radius: 1px;
}

.c0.variant-button {
  background: #1F51FF;
  border-radius: 0.8rem;
  border: 2px solid #EEEEEF;
  color: #ffffff;
  cursor: pointer;
  font-weight: 700;
  min-width: 18rem;
  padding: 1.8rem 3.2rem;
  text-shadow: 0px 1px 1px rgba(0,0,0,0.25);
  -webkit-transition: 0.2s;
  transition: 0.2s;
  white-space: nowrap;
  line-height: 1;
  font-size: 1.6rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  gap: 1rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0.variant-button:hover,
.c0.variant-button:hover:not([disabled]) {
  background: #0000FF;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0.variant-button.secondary {
  background: #ffffff;
  border: 2px solid #585858;
  border-radius: 6px;
  color: #585858;
  text-shadow: none;
}

.c0.variant-button.secondary:hover {
  background: #ffffff !important;
  border: 2px solid #000000;
  color: #000000;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0.variant-button.icon {
  background: none;
  border: none;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  min-width: 0;
  text-shadow: none;
}

.c0.variant-button.icon:hover {
  background: none;
}

.c0.variant-button.icon:focus {
  outline-color: #00594C;
}

.c0.variant-button.btn-medium {
  font-size: 1.6rem;
  min-width: 16rem;
  padding: 1.2rem 2.4rem;
}

.c0.variant-button.btn-small {
  font-size: 1.4rem;
  min-width: 15rem;
  padding: 0.8rem 2.4rem;
}

.c0.variant-button.block {
  display: block;
}

.c0.variant-button.block {
  width: 100%;
}

.c0.variant-button[disabled] {
  cursor: not-allowed !important;
  opacity: 0.5;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

<div>
  <a
    class="c0 variant-link"
  >
    Link Text
  </a>
</div>
`;
